### PR TITLE
[kuberay][autoscaler] Improve CPU, GPU, and memory detection.

### PR DESCRIPTION
**Pull request by DmitriGekhtman**
## Why are these changes needed?

After the discussion here https://github.com/ray-project/kuberay/pull/318#discussion_r909132470, I think it would be better
to have the autoscaler auto-fill Ray pod resource capacity, independently of the operator.

(This obviates the need for status.detectedRayResources or status.autoscalerHints or whatever in the RayClusterCR.)

To that end, this PR improves the autoscaler's resource detection logic.

Additions in this PR

- Autoscaler determines memory resource capacity from container config using Kubernetes `parse_resources` utility.
- Autoscaler determines GPU resource capacity from container config.
- The function used to determine CPU capacity is improved by using the Kubernetes `parse_resources` utility.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/autoscaler/_private/kuberay/autoscaling_config.py [hunk 1]
```diff
@@ -280,46 +277,53 @@ def _get_num_cpus(
...
-                        " set the `--num-gpus` rayStartParam."
-                    )
-                    _GPU_WARNING_LOGGED = True
-            break
-
+    else:
+        for key in k8s_resource_limits:
+            # e.g. nvidia.com/gpu
+            if key.endswith("/gpu"):
+                num_gpus = int(k8s_resource_limits[key])
```
Comment (sriram-anyscale) [inline python/ray/autoscaler/_private/kuberay/autoscaling_config.py:322]: Is this always going to be an int? In the other cases it is a string and you call _round_up_k8s_quantity

**Review by sriram-anyscale — APPROVED** [review]
Other than the one clarification (in a code comment), everything looks good.

**Review by pcmoritz — APPROVED** [review]
This looks like a good change to me, thanks for implementing it (it also gets rid of the possible papercut of not passing num-gpus into ray start on a GPU node, is that right?)

Let's make sure that we add an end-to-end walkthrough on how to add a GPUs to a KubeRay cluster in the Ray documentation! Similar to https://docs.ray.io/en/latest/cluster/kubernetes-gpu.html?highlight=nvidia.com%2Fgpu#gpu-usage-with-kubernetes (maybe that section just needs to be transferred to KubeRay going forward).

**Comment by DmitriGekhtman** [comment]
> (it also gets rid of the possible papercut of not passing num-gpus into ray start on a GPU node, is that right?)

Right, combined with the recent bug fix on the KubeRay side, it gets rid of the papercut.
"possible" -> "nearly unavoidable" (everyone hits the issue)

> end-to-end walkthrough on how to add a GPUs to a KubeRay cluster in the Ray documentation!

Will do.

File: python/ray/autoscaler/_private/kuberay/autoscaling_config.py (hunk 1 above)
Comment (DmitriGekhtman) [reply python/ray/autoscaler/_private/kuberay/autoscaling_config.py]: Good point
It's probably always going to be an int represented as a string.
However, to be extra safe I've switched to using `_round_up_k8s_quantity` for this as well.

**Comment by DmitriGekhtman** [comment]
Test failures unrelated.

**Pull request merged**